include/sos/barriers/WeightedTwoSidedBarrier.h

```diff
@@ -0,0 +1,148 @@
+#ifndef WEIGHTEDTWOSIDEDBARIER_H
+#define WEIGHTEDTWOSIDEDBARIER_H
+
+#include "Eigen/Eigen"
+#include "cartesian_geom/cartesian_kernel.h"
+#include <vector>
+
+template <class Point> class WeightedTwoSidedBarrier {
+
+  using NT = typename Point::FT;
+  using MT = Eigen::Matrix<NT, Eigen::Dynamic, Eigen::Dynamic>;
+  using VT = Eigen::Matrix<NT, Eigen::Dynamic, 1>;
+
+public:
+  VT lb;
+  VT ub;
+  int vdim;
+  int n;
+  std::vector<int> upperIdx;
+  std::vector<int> lowerIdx;
+  std::vector<int> freeIdx;
+  VT center;
+  const NT max_step = 1e16; // largest step size
+  VT extraHessian;
+
+  const NT inf = std::numeric_limits<NT>::infinity();
+
+  VT w;
+
+  WeightedTwoSidedBarrier(VT const &_lb, VT const &_ub, VT const &_w,
+                          int _vdim = 1) {
+    set_bound(_lb, _ub);
+    w = _w;
+    vdim = _vdim;
+    extraHessian = (1e-20) * VT::Ones(n);
+  }
+  WeightedTwoSidedBarrier() { vdim = 1; }
+
+  VT gradient(VT const &x) {
+    return w.cwiseQuotient(ub - x) - w.cwiseQuotient(x - lb);
+  }
+
+  VT hessian(VT const &x) {
+    VT d = w.cwiseQuotient((ub - x).cwiseProduct((ub - x))) +
+           w.cwiseQuotient((x - lb).cwiseProduct((x - lb)));
+    return d + extraHessian;
+  }
+  VT tensor(VT const &x) {
+    VT d = 2 * w.cwiseQuotient(
+                   ((ub - x).cwiseProduct((ub - x))).cwiseProduct((ub - x))) -
+           2 * w.cwiseQuotient(
+                   ((x - lb).cwiseProduct((x - lb))).cwiseProduct((x - lb)));
+    return d;
+  }
+  VT quadratic_form_gradient(VT const &x, VT const &u) {
+    // Output the -grad of u' (hess phi(x)) u.
+
+    return (u.cwiseProduct(u)).cwiseProduct(tensor(x));
+  }
+  NT step_size(VT const &x, VT const &v) {
+    // Output the maximum step size from x with direction v.
+
+    // check positive direction
+    VT temp = (v.array() > 0).select((ub - x).cwiseQuotient(v), max_step);
+    NT t1 = temp.minCoeff();
+
+    // check negative direction
+    temp = (v.array() < 0).select((lb - x).cwiseQuotient(v), max_step);
+    NT t2 = temp.minCoeff();
+
+    return std::min(t1, t2);
+  }
+  VT boundary_distance(VT const &x) {
+    // Output the distance of x with its closest boundary for each
+    // coordinate
+
+    return ((x - lb).cwiseMin(ub - x)).cwiseAbs();
+  }
+
+  bool feasible(VT const &x) {
+    return (x.array() > lb.array() && x.array() < ub.array()).all();
+  }
+
+  void set_bound(VT const &_lb, VT const &_ub) {
+
+    lb = _lb;
+    ub = _ub;
+    n = lb.rows();
+    extraHessian = (1e-20) * VT::Ones(n);
+    int x1 = 0, x2 = 0, x3 = 0;
+    for (int i = 0; i < n; i++) {
+      if (lb(i) == -inf) {
+        upperIdx.push_back(i);
+        x1++;
+      }
+      if (ub(i) == inf) {
+        lowerIdx.push_back(i);
+        x2++;
+      }
+      if (ub(i) == inf && lb(i) == -inf) {
+        freeIdx.push_back(i);
+        x3++;
+      }
+    }
+
+    VT c = (ub + lb) / 2;
+
+    c(lowerIdx) = lb(lowerIdx) + VT::Ones(x2, 1) * 1e6;
```

Comment:
does it makes sense to add `1e6` as a constant in the beginning?

Reply:
ok